### DISCUSSION TITLE
Vim mode `j k` for escape, keybinding context, <leader>

### DISCUSSION THREAD

**Discussion by calebmeyer** [zed-industries/zed#6661] [Feature Requests]
I put this in my keybindings:
```
[
    {
        "context": "Editor",
        "bindings": {
            "j k": "vim::NormalBefore",
            "k j": "vim::NormalBefore"
        }
    }
]
```

and now jk works for going to normal mode... but I can no longer type a j in the editor in insert mode. I think it's very possible I'm "holding it wrong" so to speak, but I'd love some guidance here. Ultimately, here's what I'd like to happen:
- Pressing `j` in insert mode inserts a literal `j` into the editor.
- Pressing any other key, or waiting say 250ms makes that `j` permanent.
- Pressing `k` instead, within the timeout, erases the `j` in the editor and puts me in normal mode. 

Other, related vim mode questions: 
- Is there a context for defining keybindings specific to one (or more) modes? E.g. I like to do `g /` for comment selected/current line in visual/normal

Answered [below](https://github.com/zed-industries/community/discussions/1026#discussioncomment-5245133) by mikayla-maki:
> you can add an `&& vim_mode == normal` to the 'context' to only fire in normal mode

- Is there any way I could use a leader? I think spacemacs finest contribution was mnemonic leader based keybindings, like `space f s` for Files: Save

Edit: I found #176 which is asking for much the same thing. Sorry if this request is in the wrong spot.

**Comment by mikayla-maki** [comment]
Hi Caleb, I think you can add an  `&& vim_mode == normal` to the 'context' field and it should only fire in normal mode :)

Our vim mode documentation does need a lot of work, thanks for highlighting the problem :)

**Reply by danielo515** [reply]
But the point is to fire it in insert mode, and still allow to type j and k separately. This is a common combo for many people that doesn't have the esc key in a nice position.

**Reply by ckipp01** [reply]
I think in this case @danielo515 you can just use `insert` instead. For example I have a similiar setup with `jj`.

```json
[
  {
    "context": "Editor && vim_mode == insert",
    "bindings": {
      "j j": "vim::NormalBefore"
    }
  }
]
```

Which seems to behave the way I'd expect.

**Reply by OlegGulevskyy** [reply]
Amazing, didn't know about "vim_mode == normal" combo, exactly what I needed!
Thanks!

**Reply by prestonButton** [reply]
This still did not work for me. I am still unable to type "j" in insert mode when I have that keybinding set. Did anyone manage to fix this problem?

Also, I prefer to map my caps lock key to `vim::NormalBefore` but I cannot figure out how to select caps lock in the `keymap.json` file. I tried "capslock" and "caps_lock" but neither of those worked.

**Reply by danielkvores** [reply]
I have the same issue, I can't type j either.

**Comment by charlesetc** [comment]
I have the same issue of wanting to type two characters (yw in my case) in insert move and moving to normal mode. Now I can't type "y" at all! Any ideas? Thank you!

**Comment by barath007183** [comment]
Hey guys this issue still persists. Has anybody found any solutions yet ?

**Reply by ch4nd4n** [reply]
In March 2025 this is what works for me. Note that this has changed a bit historically.


title keymap.json

```json
  {
    "context": "Editor && vim_mode == insert",
    "bindings": {
      "j k": ["workspace::SendKeystrokes", "escape"]
    }
  }

```

**Comment by fminkowski** [comment]
This works for me:

```json
  {
    "context": "Editor && vim_mode == insert && !menu",
    "bindings": {
      "j k": ["vim::SwitchMode", "Normal"]
    }
  }
  ```

**Reply by mrnugget** [reply]
Yup, this should work now after https://github.com/zed-industries/zed/pull/4189

**Reply by olafurpg** [reply]
This works great! I was missing this feature when I last tried Zed. I Googled `jk zed vim` and this page is now the first result :)

**Reply by nmsobri** [reply]
> This works for me:
> 
> ```json
>   {
>     "context": "Editor && vim_mode == insert && !menu",
>     "bindings": {
>       "j k": ["vim::SwitchMode", "Normal"]
>     }
>   }
> ```

but this introduce another problem, when you are in insert mode, and you want to type literal `j`, u need to wait for 1 seconds.
is there a way to configure vim timeout delay? i believe its default to 1000ms which is too long

**Reply by zerodot-sean** [reply]
I second the want to configure that delay

**Reply by mgsloan** [reply]
This is now

```
  {
    "context": "Editor && vim_mode == insert && !menu",
    "bindings": {
      "j k": "vim::SwitchToNormalMode"
    }
  }
```

**Reply by omniwrench** [reply]
> This is now
> 
> ```
>   {
>     "context": "Editor && vim_mode == insert && !menu",
>     "bindings": {
>       "j k": "vim::SwitchToNormalMode"
>     }
>   }
> ```

Is ["workspace::SendKeystrokes", "escape"] better than "vim::SwitchToNormalMode"? I think when I downloaded it previously, it had this set to send escape. Though I did download the latest version again, and now it's defaulting to "vim::SwitchToNormalMode". Also the docs seems to still say "vim::NormalBefore".

**Reply by mgsloan** [reply]
@omniwrench Thanks for noticing this!  Initial keymap is fixed to use NormalBefore in #34912

So indeed if you want "j k" to match the behavior of escape then this should typically be

```
  {
    "context": "Editor && vim_mode == insert",
    "bindings": {
      "j k": "vim::NormalBefore"
    }
  }
```

This also now removes the `!menu` - see https://github.com/zed-industries/zed/pull/34912#issuecomment-3108802582

**Comment by mrnugget** [comment]
Adding it to the docs: https://github.com/zed-industries/zed/pull/11150

**Comment by TrevorHinesley** [comment]
@mrnugget here's an example where, apparently due to the autosuggestion list being visible, it disables the `jk` binding until that menu is closed. Is there any way to have this work even in this scenario?

https://github.com/user-attachments/assets/8701d59c-24a1-47eb-b9ba-7251c2337fbd

**Reply by mrnugget** [reply]
Have you tried removing the `!menu` from the keybinding context?

```json
{
    "context": "Editor && vim_mode == insert",
    "bindings": {
      "j k": ["vim::SwitchMode", "Normal"]
    }
}
```

**Reply by itsmejoeyb** [reply]
Removing !menu kind of works. It will exit insert mode but the menu stays open and hjkl won't move your cursor unless you click away from the menu to close it.

**Reply by Or-Gold** [reply]
@itsmejoeyb this issue is not recent - the correct binding is now [in the docs](https://zed.dev/docs/vim#custom-key-bindings) and works in autocomplete aswell

```json
{
    "context": "vim_mode == insert",
    "bindings": {
      "j k": "vim::NormalBefore"
    }
},
```

**Reply by mapsgeek** [reply]
the menu stays open, is there a way to combine exiting the mode and close the menu at the same time?
